### PR TITLE
🌱 Improve KCP remediation re-entrancy

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR improves KCP remediation in case a controller crash happens in between creating the new machine and patching KCP.

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes-sigs/cluster-api/issues/8188

/area control-plane

### CODE REVIEW

**Review by chrischdi — COMMENTED** [review]
/lgtm

**Review by sbueringer — COMMENTED** [review]
Thx!

Just one nit

File: controlplane/kubeadm/internal/controllers/remediation.go [hunk 1]
```diff
@@ -99,9 +99,29 @@ func (r *KubeadmControlPlaneReconciler) reconcileUnhealthyMachines(ctx context.C
 	// Returns if another remediation is in progress but the new Machine is not yet created.
 	// Note: This condition is checked after we check for unhealthy Machines and if machineToBeRemediated
 	// is being deleted to avoid unnecessary logs if no further remediation should be done.
-	if _, ok := controlPlane.KCP.Annotations[controlplanev1.RemediationInProgressAnnotation]; ok {
-		log.Info("Another remediation is already in progress. Skipping remediation.")
-		return ctrl.Result{}, nil
+	if v, ok := controlPlane.KCP.Annotations[controlplanev1.RemediationInProgressAnnotation]; ok {
+		// Check if the annotation is stale; this might happen in case there is a crash in the controller in between
+		// when the annotation is removed from KCP after the replacement machine has been created, and the defer patch at the end
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply controlplane/kubeadm/internal/controllers/remediation.go:104]: @fabriziopandini ^^

Otherwise happy to merge

**Comment by sbueringer** [comment]
/lgtm
/approve

Thx!

**Pull request merged**